json/src/fs2/data/json/selector/dsl.scala

```diff
@@ -0,0 +1,175 @@
+    * The resulting selector is strict and will fail if the element is not an object.
+    */
+  def fields(f: String, fs: String*): NamesBuilder[Optional, Strict] =
+    NamesBuilder[Optional, Strict](NamePredicate.Several(fs.toSet + f), true, false, this)
+
+  /** Compiles the current builder into its final selector representation. */
+  def compile: Selector
+
+}
+
+case object RootBuilder extends SelectorBuilder[NotApplicable, NotApplicable] {
+  def compile: Selector = Selector.ThisSelector
+}
+
+case class IteratorBuilder[S](strict: Boolean, parent: SelectorBuilder[_, _])
+    extends SelectorBuilder[NotApplicable, S] {
+  def compile: Selector = Selector.PipeSelector.from(parent.compile, Selector.IteratorSelector(strict))
+}
+
+object IteratorBuilder {
+
+  implicit def LenientableIterator: Lenientable.Aux[IteratorBuilder[Strict], IteratorBuilder[Lenient], NotApplicable] =
+    new Lenientable[IteratorBuilder[Strict], NotApplicable] {
+    type Out = IteratorBuilder[Lenient]
+    def makeLenient(builder: IteratorBuilder[Strict]): IteratorBuilder[Lenient] =
+      builder.copy(strict = false)
+  }
+
+}
+
+case class IndicesBuilder[S](predicate: IndexPredicate, strict: Boolean, parent: SelectorBuilder[_, _])
+    extends SelectorBuilder[NotApplicable, S] {
+  def compile: Selector = Selector.PipeSelector.from(parent.compile, Selector.IndexSelector(predicate, strict))
+}
+
+object IndicesBuilder {
+
+  implicit def LenientableIndices: Lenientable.Aux[IndicesBuilder[Strict], IndicesBuilder[Lenient], NotApplicable] =
+    new Lenientable[IndicesBuilder[Strict], NotApplicable] {
+      type Out = IndicesBuilder[Lenient]
+      def makeLenient(builder: IndicesBuilder[Strict]): IndicesBuilder[Lenient] =
+        builder.copy(strict = false)
+    }
+
+}
+
+case class NamesBuilder[M, S](predicate: NamePredicate,
+                              strict: Boolean,
+                              mandatory: Boolean,
+                              parent: SelectorBuilder[_, _])
+    extends SelectorBuilder[M, S] {
+  def compile: Selector =
+    Selector.PipeSelector.from(parent.compile, Selector.NameSelector(predicate, strict, mandatory))
+}
+
+object NamesBuilder {
+
+  implicit def LenientableNames[M]: Lenientable.Aux[NamesBuilder[M, Strict], NamesBuilder[M, Lenient], M] =
+    new Lenientable[NamesBuilder[M, Strict], M] {
+      type Out = NamesBuilder[M, Lenient]
+      def makeLenient(builder: NamesBuilder[M, Strict]): NamesBuilder[M, Lenient] =
+        builder.copy(strict = false)
+    }
+
+  implicit def MandatoriableNames[S]: Mandatoriable.Aux[NamesBuilder[Optional, S], NamesBuilder[Mandatory, S], S] =
+    new Mandatoriable[NamesBuilder[Optional, S], S] {
+      type Out = NamesBuilder[Mandatory, S]
+      def makeMandatory(builder: NamesBuilder[Optional, S]): NamesBuilder[Mandatory, S] =
+        builder.copy(mandatory = true)
+    }
+
+}
+
+class NotApplicable private {}
```

Comment:
Let's add a small comment about the meaning of each of these markers? Makes reading this file a lot easier if you don't have to figure out by yourself what means what and what is the dual of which other marker.

Reply:
Yep.